jetstream_pt/third_party/mistral/model_original.py

```diff
@@ -0,0 +1,281 @@
+    self.block_sparse_moe = MOEFeedForward(config)
+    self.ffn_norm = RMSNorm(config.dim, config.norm_eps)
+    self.attention_norm = RMSNorm(config.dim, config.norm_eps)
+
+  def forward(
+      self, x: Tensor, input_pos: Tensor, freqs_cis: Tensor, mask: Tensor
+  ) -> Tensor:
+    h = x + self.attention(self.attention_norm(x), freqs_cis, mask, input_pos)
+    out = h + self.block_sparse_moe(self.ffn_norm(h))
+    return out
+
+
+class Attention(nn.Module):
+
+  def __init__(self, config: ModelArgs):
+    super().__init__()
+    assert config.dim % config.n_head == 0
+
+    total_head_dim = (
+        config.n_head + 2 * config.n_local_heads
+    ) * config.head_dim
+    # key, query, value projections for all heads, but in a batch
+    self.wqkv = nn.Linear(config.dim, total_head_dim, bias=False)
+    self.wo = nn.Linear(config.dim, config.dim, bias=False)
+    self.kv_cache = None
+
+    self.n_head = config.n_head
+    self.head_dim = config.head_dim
+    self.n_local_heads = config.n_local_heads
+    self.dim = config.dim
+    self._register_load_state_dict_pre_hook(self.load_hook)
+
+  def load_hook(self, state_dict, prefix, *args):
+    if prefix + "wq.weight" in state_dict:
+      wq = state_dict.pop(prefix + "wq.weight")
+      wk = state_dict.pop(prefix + "wk.weight")
+      wv = state_dict.pop(prefix + "wv.weight")
+      state_dict[prefix + "wqkv.weight"] = torch.cat([wq, wk, wv])
+
+  def forward(
+      self,
+      x: Tensor,
+      freqs_cis: Tensor,
+      mask: Tensor,
+      input_pos: Optional[Tensor] = None,
+  ) -> Tensor:
+    bsz, seqlen, _ = x.shape
+
+    kv_size = self.n_local_heads * self.head_dim
+    q, k, v = self.wqkv(x).split([self.dim, kv_size, kv_size], dim=-1)
+
+    q = q.view(bsz, seqlen, self.n_head, self.head_dim)
+    k = k.view(bsz, seqlen, self.n_local_heads, self.head_dim)
+    v = v.view(bsz, seqlen, self.n_local_heads, self.head_dim)
+
+    q = apply_rotary_emb(q, freqs_cis)
+    k = apply_rotary_emb(k, freqs_cis)
+
+    q, k, v = map(lambda x: x.transpose(1, 2), (q, k, v))
+
+    if self.kv_cache is not None:
+      k, v = self.kv_cache.update(input_pos, k, v)
+
+    k = k.repeat_interleave(self.n_head // self.n_local_heads, dim=1)
+    v = v.repeat_interleave(self.n_head // self.n_local_heads, dim=1)
+    y = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0)
+
+    y = y.transpose(1, 2).contiguous().view(bsz, seqlen, self.dim)
+
+    y = self.wo(y)
+    return y
+
+
+class ConditionalFeedForward(nn.Module):
+
+  def __init__(self, config):
+    super().__init__()
+    self.w1 = nn.Parameter(
+        torch.empty(config.num_experts, config.intermediate_size, config.dim)
+    )
+    self.w2 = nn.Parameter(
+        torch.empty(config.num_experts, config.dim, config.intermediate_size)
+    )
+    self.w3 = nn.Parameter(
+        torch.empty(config.num_experts, config.intermediate_size, config.dim)
+    )
+
+  def forward(self, x: Tensor, expert_indices: Tensor) -> Tensor:
```

Comment:
I had a change to use different logic for longer seqlen and i pushed to your branch, is that lost from merging?

Reply:
also the quantized change

Reply:
This is the original model. Your changes are in model.py